airflow/models.py

```diff
@@ -3055,6 +3055,37 @@ def latest_execution_date(self):
         session.close()
         return execution_date
 
+    @property
+    def next_run_date(self):
+        """
+        Returns the next run date for which the dag will be scheduled
+        """
+        next_run_date = None
+        if not self.latest_execution_date:
+            # First run
+            task_start_dates = [t.start_date for t in self.tasks]
+            if task_start_dates:
+                next_run_date = self.normalize_schedule(min(task_start_dates))
+        else:
+            next_run_date = self.following_schedule(self.latest_execution_date)
+        return next_run_date
+
+    @property
+    def next_execution_date(self):
+        """
+        Returns the next execution date at which the dag will be scheduled by
```

Comment:
It's not clear to me how these two methods differ. What's the intent behind these two methods?

Reply:
I don't think there is a distinction between execution date an run date - it's not something I'm familiar with and to my mind they were two names for the same concept.

I think the overall logic is roughly this:

- If there are any previous dag runs for for the dag, then the next date is `following_schedule(latest_dag_run.execution_date)` (possibly filtering out "adhoc" or backfilled dagruns. Not sure)
- Otherwise it is start date

I think it's that simple. I'm not 100% familiar with all the details of the scheduler so I'm not 100% sure. This is what the `normalie_scheduled` function does.

This bit of jobs.py seems to be relevant https://github.com/apache/incubator-airflow/blob/e53182cf83501a789fbdd74ee472d29052a5e321/airflow/jobs.py#L840-L851

Reply:
This is exactly the logic I've followed and you can see that my proposed implementation is based on the scheduler's jobs.py code indeed

I can shrink the next_run_date function into one to make it simple tho indeed, gonna update

@XD-DENG as you see the scheduler does otherwise and having None as a result looks strange to me since there's always a **period end** at which the scheduler itself will execute the DAG (if it has a start_date that is).

So it's not because no previous execution has happened that there won't be any right. And the scheduler code above shows how it does it.

Still, @XD-DENG I think your link has a nice example and I'll make sure to validate the fact that this PR behaves exactly as the documentation says. Sound good to you?